BnbnavNetClient/Views/MapView.axaml.cs

```diff
@@ -90,25 +104,77 @@ protected override void OnInitialized()
 
                 _toScreenMtx = matrix;
                 _toWorldMtx = matrix.Invert();
+                
+                UpdateDrawnItems();
             }));
-
-
         MapViewModel
             .WhenAnyPropertyChanged()
             .Subscribe(Observer.Create<MapViewModel?>(_ => { InvalidateVisual(); }));
     }
 
     static readonly double LandmarkSize = 10;
+    static readonly double NodeSize = 14;
     readonly Dictionary<string, SKSvg> _svgCache = new();
 
-    public override void Render(DrawingContext context)
+    IReadOnlyList<(Point, Point, Edge)> _drawnEdges = new List<(Point, Point, Edge)>();
+    IReadOnlyList<(Rect, Landmark)> _drawnLandmarks = new List<(Rect, Landmark)>();
+    IReadOnlyList<(Rect, Node)> _drawnNodes = new List<(Rect, Node)>();
+
+    private IEnumerable<MapItem> HitTest(Point point)
+    {
+        foreach (var (rect, landmark) in _drawnLandmarks)
+        {
+            if (rect.Contains(point)) yield return landmark;
+        }
+        
+        foreach (var (rect, node) in _drawnNodes)
+        {
+            if (rect.Contains(point)) yield return node;
+        }
+    }
+
+    private void UpdateDrawnItems()
     {
         var mapService = MapViewModel.MapService;
         var scale = MapViewModel.Scale;
 
+        _drawnEdges = mapService.Edges.Values.Select(edge =>
+        {
+            var fromEdge = mapService.Nodes[edge.From.Id];
+            var from = ToScreen(new(fromEdge.X, fromEdge.Z));
+            var toEdge = mapService.Nodes[edge.To.Id];
+            var to = ToScreen(new (toEdge.X, toEdge.Z));
+            return (from, to, edge);
+        }).Where(edge => LineIntersects(edge.from, edge.to, Bounds)).ToList();
+
+        _drawnLandmarks = mapService.Landmarks.Values.Select(landmark =>
+        {
+            var pos = ToScreen(new(landmark.Node.X, landmark.Node.Z));
+            var rect = new Rect(
+                pos.X - LandmarkSize * scale / 2,
+                pos.Y - LandmarkSize * scale / 2,
+                LandmarkSize * scale, LandmarkSize * scale);
+            return (rect, landmark);
+        }).Where(landmark => Bounds.Intersects(landmark.rect)).ToList().AsReadOnly();
+
+        _drawnNodes = mapService.Nodes.Values.Select(node =>
+        {
+            var pos = ToScreen(new(node.X, node.Z));
+            var rect = new Rect(
+                pos.X - NodeSize / 2, 
+                pos.Y - NodeSize / 2,
+                NodeSize, NodeSize);
+            return (rect, node);
+        }).Where(node => Bounds.Intersects(node.rect)).ToList().AsReadOnly();
```

Comment:
Please note that there are performance implications to using LINQ and, if my memory serves me right, read-only collections too. Maybe we should reconsider and/or rewrite

Reply:
@reflectronic your thoughts?

Reply:
Performance seems to be okay so far - if required we can come back and look at this but I don't think it's an issue right now